File: js/src/discovery.mjs
import { readFile, readdir, realpath } from 'node:fs/promises';
import path from 'node:path';

import { REGISTRIES } from './model.mjs';

const IGNORED_DIRECTORIES = new Set([
  '.git',
  '.package-registry-manager',
  '.venv',
  'node_modules',
  'target',
  'vendor',
]);

const MANIFEST_NAMES = new Set([
  'package.json',
  'Cargo.toml',
  'pyproject.toml',
  'setup.py',
  'go.mod',
  'pom.xml',
  'build.gradle',
  'build.gradle.kts',
  'composer.json',
]);

export async function inspectRepository(repository) {
  const root = await realpath(repository);
  const manifests = [];
  await collectManifests(root, manifests);
  manifests.sort();
  const preferredManifests = manifests.filter(
    (manifest) =>
      path.basename(manifest) !== 'setup.py' ||
      !manifests.includes(path.join(path.dirname(manifest), 'pyproject.toml'))
  );
  const packages = (
    await Promise.all(
      preferredManifests.map(async (manifest) =>
        parseManifest(manifest, relativePath(root, manifest))
      )
    )
  )
    .filter(Boolean)
    .sort((left, right) => {
      const registryOrder =
        REGISTRIES.indexOf(left.registry) - REGISTRIES.indexOf(right.registry);
      return (
        registryOrder ||
        [left.manifest, left.name]
          .join('\0')
          .localeCompare([right.manifest, right.name].join('\0'))
      );
    });
  const { github_owner, github_repository } = await githubCoordinates(
    root,
    preferredManifests
  );

  return {
    schema_version: 1,
    repository: {
      root,
      github_owner,
      github_repository,
      release_workflow: await releaseWorkflow(root),
    },
    packages,
  };
}

async function collectManifests(directory, manifests) {
  const entries = await readdir(directory, { withFileTypes: true });
  entries.sort((left, right) => left.name.localeCompare(right.name));
  for (const entry of entries) {
    const item = path.join(directory, entry.name);
    if (entry.isDirectory() && !IGNORED_DIRECTORIES.has(entry.name)) {
      await collectManifests(item, manifests);
    } else if (
      entry.isFile() &&
      (MANIFEST_NAMES.has(entry.name) || /\.(?:csproj|fsproj|vbproj)$/.test(entry.name))
    ) {
      manifests.push(item);
    }
  }
}

async function parseManifest(manifestPath, manifest) {
  const contents = await readFile(manifestPath, 'utf8');
  const filename = path.basename(manifestPath);
  switch (filename) {
    case 'package.json':
      return parseNpm(contents, manifest);
    case 'Cargo.toml':
      return parseCargo(contents, manifest);
    case 'pyproject.toml':
      return parsePyproject(contents, manifest);
    case 'setup.py':
      return packageInfo(
        'pypi',
        capture(contents, /\bname\s*=\s*['"]([^'"]+)/) ??
          'unknown-python-package',
        capture(contents, /\bversion\s*=\s*['"]([^'"]+)/),
        manifest
      );
    case 'go.mod':
      return packageInfo(
        'go-modules',
        capture(contents, /^\s*module\s+(\S+)/m) ?? 'unknown-go-module',
        null,
        manifest
      );
    case 'pom.xml':
      return parseMaven(contents, manifest);
    case 'build.gradle':
    case 'build.gradle.kts':
      return parseGradle(contents, manifest);
    case 'composer.json':
      return parseComposer(contents, manifest);
    default:
      if (/\.(?:csproj|fsproj|vbproj)$/.test(filename)) {
        return parseDotnet(contents, manifestPath, manifest);
      }
      return null;
  }
}

function parseNpm(contents, manifest) {
  const metadata = parseJson(contents, manifest);
  if (!metadata.name) return null;
  const publishable = metadata.private !== true;
  return packageInfo(
    'npm',
    metadata.name,
    metadata.version ?? null,
    manifest,
    publishable,
    publishable ? [] : ['package.json marks this package as private']
  );
}

function parseCargo(contents, manifest) {
  const section = tomlSection(contents, 'package');
  if (!section) return null;
  const name = tomlString(section, 'name');
  if (!name) return null;
  const publishable = !/^\s*publish\s*=\s*false\s*$/m.test(section);
  return packageInfo(
    'crates-io',
    name,
    tomlString(section, 'version'),
    manifest,
    publishable,
    publishable ? [] : ['Cargo.toml disables publishing']
  );
}

function parsePyproject(contents, manifest) {
  const section =
    tomlSection(contents, 'project') ?? tomlSection(contents, 'tool.poetry');
  const name = section && tomlString(section, 'name');
  if (!name) return null;
  return packageInfo('pypi', name, tomlString(section, 'version'), manifest);
}

function parseDotnet(contents, manifestPath, manifest) {
  return packageInfo(
    'nuget',
    xmlTag(contents, 'PackageId') ??
      xmlTag(contents, 'AssemblyName') ??
      path.basename(manifestPath, path.extname(manifestPath)),
    xmlTag(contents, 'PackageVersion') ?? xmlTag(contents, 'Version'),
    manifest
  );
}

function parseMaven(contents, manifest) {
  const artifact = xmlTag(contents, 'artifactId') ?? 'unknown-maven-artifact';
  const group = xmlTag(contents, 'groupId');
  return packageInfo(
    'maven-central',
    group ? `${group}:${artifact}` : artifact,
    xmlTag(contents, 'version'),
    manifest
  );
}

function parseGradle(contents, manifest) {
  const group = capture(contents, /^\s*group\s*=\s*['"]([^'"]+)/m);
  const artifact =
    capture(
      contents,
      /^\s*(?:archivesBaseName|rootProject\.name)\s*=\s*['"]([^'"]+)/m
    ) ?? 'gradle-project';
  return packageInfo(
    'maven-central',
    group ? `${group}:${artifact}` : artifact,
    capture(contents, /^\s*version\s*=\s*['"]([^'"]+)/m),
    manifest
  );
}

function parseComposer(contents, manifest) {
  const metadata = parseJson(contents, manifest);
  if (!metadata.name) return null;
  return packageInfo(
    'packagist',
    metadata.name,
    metadata.version ?? null,
    manifest
  );
}

function packageInfo(
  registry,
  name,
  version,
  manifest,
  publishable = true,
  problems = []
) {
  const result = { registry, name, version: version ?? null, manifest, publishable };
  if (problems.length > 0) result.problems = problems;
  return result;
}

function parseJson(contents, manifest) {
  try {
    return JSON.parse(contents);
  } catch (error) {
    throw new Error(`invalid JSON in ${manifest}: ${error.message}`, { cause: error });
  }
}

function tomlSection(contents, name) {
  const escaped = name.replaceAll('.', '\\.');
  return capture(
    contents,
    new RegExp(
      `^\\[${escaped}\\]\\s*$([\\s\\S]*?)(?=^\\[|(?![\\s\\S]))`,
      'm'
    )
  );
}

function tomlString(section, key) {
  return capture(section, new RegExp(`^\\s*${key}\\s*=\\s*['\"]([^'\"]+)`, 'm'));
}

function xmlTag(contents, tag) {
  return capture(contents, new RegExp(`<${tag}(?:\\s[^>]*)?>\\s*([^<]+?)\\s*</${tag}>`, 's'));
}

function capture(contents, expression) {
  return expression.exec(contents)?.[1] ?? null;
}

function relativePath(root, item) {
  return path.relative(root, item).split(path.sep).join('/');
}

async function githubCoordinates(root, manifests) {
  let contents;
  try {
    contents = await readFile(path.join(root, '.git/config'), 'utf8');
  } catch {
    contents = '';
  }
  const remote = capture(contents, /^\s*url\s*=\s*(\S+)/m);
  const remoteCoordinates = parseGithubUrl(remote);
  if (remoteCoordinates) return remoteCoordinates;

  for (const manifest of manifests) {
    const filename = path.basename(manifest);
    contents = await readFile(manifest, 'utf8');
    let repository;
    if (filename === 'package.json') {
      const metadata = parseJson(contents, relativePath(root, manifest));
      repository =
        typeof metadata.repository === 'string'
          ? metadata.repository
          : metadata.repository?.url;
    } else if (filename === 'Cargo.toml') {
      const packageSection = tomlSection(contents, 'package');
      repository = packageSection && tomlString(packageSection, 'repository');
    }
    const coordinates = parseGithubUrl(repository);
    if (coordinates) return coordinates;
  }
  return { github_owner: null, github_repository: null };
}

function parseGithubUrl(remote) {
  const normalized = remote
    ?.replace(/\.git$/, '')
    .replace(/^git\+/, '')
    .replace('git@github.com:', 'https://github.com/')
    .replace('ssh://git@github.com/', 'https://github.com/');
  if (!normalized?.startsWith('https://github.com/')) {
    return null;
  }
  const [github_owner, github_repository] = normalized
    .slice('https://github.com/'.length)
    .split('/');
  return { github_owner: github_owner ?? null, github_repository: github_repository ?? null };
}

async function releaseWorkflow(root) {
  const directory = path.join(root, '.github/workflows');
  let entries;
  try {
    entries = await readdir(directory, { withFileTypes: true });
  } catch {
    return null;
  }
  entries.sort((left, right) => left.name.localeCompare(right.name));
  let fallback = null;
  for (const entry of entries) {
    if (!entry.isFile() || !/\.ya?ml$/.test(entry.name)) continue;
    const contents = await readFile(path.join(directory, entry.name), 'utf8');
    if (contents.includes('npm publish') || contents.includes('npm stage publish')) {
      return entry.name;
    }
    if (!fallback && entry.name.includes('release')) fallback = entry.name;
  }
  return fallback;
}

File: src/discovery.rs
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};
use regex::Regex;
use serde_json::Value as JsonValue;
use toml::Value as TomlValue;

use crate::model::{Inspection, Package, Registry, RepositoryInfo};

const IGNORED_DIRECTORIES: &[&str] = &[
    ".git",
    ".package-registry-manager",
    ".venv",
    "node_modules",
    "target",
    "vendor",
];

/// Inspect a repository without running package-manager commands.
pub fn inspect_repository(root: &Path) -> Result<Inspection> {
    let root = root
        .canonicalize()
        .with_context(|| format!("repository does not exist: {}", root.display()))?;
    let mut manifests = Vec::new();
    collect_manifests(&root, &mut manifests)?;
    manifests.sort();
    manifests.retain(|path| {
        path.file_name().and_then(|name| name.to_str()) != Some("setup.py")
            || !path.with_file_name("pyproject.toml").is_file()
    });

    let mut packages = manifests
        .iter()
        .filter_map(|path| parse_manifest(path, &root).transpose())
        .collect::<Result<Vec<_>>>()?;
    packages.sort_by(|left, right| {
        (left.registry, &left.manifest, &left.name).cmp(&(
            right.registry,
            &right.manifest,
            &right.name,
        ))
    });

    let (github_owner, github_repository) = github_coordinates(&root, &manifests)?;
    Ok(Inspection {
        schema_version: 1,
        repository: RepositoryInfo {
            root: root.to_string_lossy().into_owned(),
            github_owner,
            github_repository,
            release_workflow: release_workflow(&root)?,
        },
        packages,
    })
}

fn collect_manifests(directory: &Path, manifests: &mut Vec<PathBuf>) -> Result<()> {
    let mut entries = fs::read_dir(directory)
        .with_context(|| format!("cannot read {}", directory.display()))?
        .collect::<std::io::Result<Vec<_>>>()?;
    entries.sort_by_key(std::fs::DirEntry::file_name);

    for entry in entries {
        let path = entry.path();
        if path.is_dir() {
            if !IGNORED_DIRECTORIES.contains(&entry.file_name().to_string_lossy().as_ref()) {
                collect_manifests(&path, manifests)?;
            }
        } else if is_manifest(&path) {
            manifests.push(path);
        }
    }
    Ok(())
}

fn is_manifest(path: &Path) -> bool {
    let Some(name) = path.file_name().and_then(|name| name.to_str()) else {
        return false;
    };
    matches!(
        name,
        "package.json"
            | "Cargo.toml"
            | "pyproject.toml"
            | "setup.py"
            | "go.mod"
            | "pom.xml"
            | "build.gradle"
            | "build.gradle.kts"
            | "composer.json"
    ) || matches!(
        path.extension().and_then(|extension| extension.to_str()),
        Some("csproj" | "fsproj" | "vbproj")
    )
}

fn parse_manifest(path: &Path, root: &Path) -> Result<Option<Package>> {
    let relative = relative_path(path, root);
    let contents = fs::read_to_string(path)
        .with_context(|| format!("cannot read manifest {}", path.display()))?;
    let name = path
        .file_name()
        .and_then(|name| name.to_str())
        .unwrap_or_default();

    match name {
        "package.json" => parse_npm(&contents, relative),
        "Cargo.toml" => parse_cargo(&contents, relative),
        "pyproject.toml" => parse_pyproject(&contents, relative),
        "setup.py" => Ok(Some(parse_setup_py(&contents, relative))),
        "go.mod" => Ok(Some(parse_go(&contents, relative))),
        "pom.xml" => Ok(Some(parse_maven(&contents, relative))),
        "build.gradle" | "build.gradle.kts" => Ok(Some(parse_gradle(&contents, relative))),
        "composer.json" => parse_composer(&contents, relative),
        _ if matches!(
            path.extension().and_then(|extension| extension.to_str()),
            Some("csproj" | "fsproj" | "vbproj")
        ) =>
        {
            Ok(Some(parse_dotnet(&contents, path, relative)))
        }
        _ => Ok(None),
    }
}

fn parse_npm(contents: &str, manifest: String) -> Result<Option<Package>> {
    let value: JsonValue =
        serde_json::from_str(contents).with_context(|| format!("invalid JSON in {manifest}"))?;
    let Some(name) = value.get("name").and_then(JsonValue::as_str) else {
        return Ok(None);
    };
    let publishable = !value
        .get("private")
        .and_then(JsonValue::as_bool)
        .unwrap_or(false);
    Ok(Some(Package {
        registry: Registry::Npm,
        name: name.to_owned(),
        version: json_string(&value, "version"),
        manifest,
        publishable,
        problems: if publishable {
            Vec::new()
        } else {
            vec!["package.json marks this package as private".to_owned()]
        },
    }))
}

fn parse_cargo(contents: &str, manifest: String) -> Result<Option<Package>> {
    let value: TomlValue =
        toml::from_str(contents).with_context(|| format!("invalid TOML in {manifest}"))?;
    let Some(package) = value.get("package") else {
        return Ok(None);
    };
    let Some(name) = package.get("name").and_then(TomlValue::as_str) else {
        return Ok(None);
    };
    let publishable = !matches!(package.get("publish"), Some(TomlValue::Boolean(false)));
    Ok(Some(Package {
        registry: Registry::CratesIo,
        name: name.to_owned(),
        version: package
            .get("version")
            .and_then(TomlValue::as_str)
            .map(str::to_owned),
        manifest,
        publishable,
        problems: if publishable {
            Vec::new()
        } else {
            vec!["Cargo.toml disables publishing".to_owned()]
        },
    }))
}

fn parse_pyproject(contents: &str, manifest: String) -> Result<Option<Package>> {
    let value: TomlValue =
        toml::from_str(contents).with_context(|| format!("invalid TOML in {manifest}"))?;
    let metadata = value
        .get("project")
        .or_else(|| value.get("tool")?.get("poetry"));
    let Some(metadata) = metadata else {
        return Ok(None);
    };
    let Some(name) = metadata.get("name").and_then(TomlValue::as_str) else {
        return Ok(None);
    };
    Ok(Some(Package {
        registry: Registry::PyPi,
        name: name.to_owned(),
        version: metadata
            .get("version")
            .and_then(TomlValue::as_str)
            .map(str::to_owned),
        manifest,
        publishable: true,
        problems: Vec::new(),
    }))
}

fn parse_setup_py(contents: &str, manifest: String) -> Package {
    Package {
        registry: Registry::PyPi,
        name: regex_capture(contents, r#"(?m)\bname\s*=\s*['\"]([^'\"]+)"#)
            .unwrap_or_else(|| "unknown-python-package".to_owned()),
        version: regex_capture(contents, r#"(?m)\bversion\s*=\s*['\"]([^'\"]+)"#),
        manifest,
        publishable: true,
        problems: Vec::new(),
    }
}

fn parse_go(contents: &str, manifest: String) -> Package {
    let name = contents
        .lines()
        .find_map(|line| line.trim().strip_prefix("module "))
        .map_or("unknown-go-module", str::trim)
        .to_owned();
    Package {
        registry: Registry::GoModules,
        name,
        version: None,
        manifest,
        publishable: true,
        problems: Vec::new(),
    }
}

fn parse_dotnet(contents: &str, path: &Path, manifest: String) -> Package {
    let fallback = path
        .file_stem()
        .and_then(|stem| stem.to_str())
        .unwrap_or("unknown-dotnet-package")
        .to_owned();
    Package {
        registry: Registry::NuGet,
        name: xml_tag(contents, "PackageId")
            .or_else(|| xml_tag(contents, "AssemblyName"))
            .unwrap_or(fallback),
        version: xml_tag(contents, "PackageVersion").or_else(|| xml_tag(contents, "Version")),
        manifest,
        publishable: true,
        problems: Vec::new(),
    }
}

fn parse_maven(contents: &str, manifest: String) -> Package {
    let artifact =
        xml_tag(contents, "artifactId").unwrap_or_else(|| "unknown-maven-artifact".to_owned());
    let group = xml_tag(contents, "groupId");
    Package {
        registry: Registry::MavenCentral,
        name: group.map_or_else(|| artifact.clone(), |group| format!("{group}:{artifact}")),
        version: xml_tag(contents, "version"),
        manifest,
        publishable: true,
        problems: Vec::new(),
    }
}

fn parse_gradle(contents: &str, manifest: String) -> Package {
    let group = regex_capture(contents, r#"(?m)^\s*group\s*=\s*['\"]([^'\"]+)"#);
    let artifact = regex_capture(
        contents,
        r#"(?m)^\s*(?:archivesBaseName|rootProject\.name)\s*=\s*['\"]([^'\"]+)"#,
    )
    .unwrap_or_else(|| "gradle-project".to_owned());
    Package {
        registry: Registry::MavenCentral,
        name: group.map_or_else(|| artifact.clone(), |group| format!("{group}:{artifact}")),
        version: regex_capture(contents, r#"(?m)^\s*version\s*=\s*['\"]([^'\"]+)"#),
        manifest,
        publishable: true,
        problems: Vec::new(),
    }
}

fn parse_composer(contents: &str, manifest: String) -> Result<Option<Package>> {
    let value: JsonValue =
        serde_json::from_str(contents).with_context(|| format!("invalid JSON in {manifest}"))?;
    let Some(name) = value.get("name").and_then(JsonValue::as_str) else {
        return Ok(None);
    };
    Ok(Some(Package {
        registry: Registry::Packagist,
        name: name.to_owned(),
        version: json_string(&value, "version"),
        manifest,
        publishable: true,
        problems: Vec::new(),
    }))
}

fn json_string(value: &JsonValue, key: &str) -> Option<String> {
    value
        .get(key)
        .and_then(JsonValue::as_str)
        .map(str::to_owned)
}

fn regex_capture(contents: &str, pattern: &str) -> Option<String> {
    Regex::new(pattern)
        .expect("static regular expression must compile")
        .captures(contents)
        .and_then(|captures| captures.get(1))
        .map(|capture| capture.as_str().to_owned())
}

fn xml_tag(contents: &str, tag: &str) -> Option<String> {
    regex_capture(
        contents,
        &format!(r"(?s)<{tag}(?:\s[^>]*)?>\s*([^<]+?)\s*</{tag}>"),
    )
}

fn relative_path(path: &Path, root: &Path) -> String {
    path.strip_prefix(root)
        .unwrap_or(path)
        .components()
        .map(|component| component.as_os_str().to_string_lossy())
        .collect::<Vec<_>>()
        .join("/")
}

fn github_coordinates(
    root: &Path,
    manifests: &[PathBuf],
) -> Result<(Option<String>, Option<String>)> {
    let config_path = root.join(".git/config");
    if let Ok(contents) = fs::read_to_string(&config_path) {
        if let Some(remote) = Regex::new(r"(?m)^\s*url\s*=\s*(\S+)")?
            .captures(&contents)
            .and_then(|captures| captures.get(1))
            .map(|capture| capture.as_str())
        {
            if let Some(coordinates) = parse_github_url(remote) {
                return Ok(coordinates);
            }
        }
    }

    for manifest in manifests {
        let contents = fs::read_to_string(manifest)?;
        let repository = match manifest.file_name().and_then(|name| name.to_str()) {
            Some("package.json") => {
                let value: JsonValue = serde_json::from_str(&contents)?;
                value
                    .get("repository")
                    .and_then(|repository| {
                        repository
                            .as_str()
                            .or_else(|| repository.get("url").and_then(JsonValue::as_str))
                    })
                    .map(str::to_owned)
            }
            Some("Cargo.toml") => {
                let value: TomlValue = toml::from_str(&contents)?;
                value
                    .get("package")
                    .and_then(|package| package.get("repository"))
                    .and_then(TomlValue::as_str)
                    .map(str::to_owned)
            }
            _ => None,
        };
        if let Some(coordinates) = repository.as_deref().and_then(parse_github_url) {
            return Ok(coordinates);
        }
    }
    Ok((None, None))
}

fn parse_github_url(remote: &str) -> Option<(Option<String>, Option<String>)> {
    let normalized = remote
        .trim_end_matches(".git")
        .trim_start_matches("git+")
        .replace("git@github.com:", "https://github.com/")
        .replace("ssh://git@github.com/", "https://github.com/");
    let path = normalized.strip_prefix("https://github.com/")?;
    let mut parts = path.split('/');
    Some((
        parts.next().map(str::to_owned),
        parts.next().map(str::to_owned),
    ))
}

fn release_workflow(root: &Path) -> Result<Option<String>> {
    let directory = root.join(".github/workflows");
    let Ok(entries) = fs::read_dir(directory) else {
        return Ok(None);
    };
    let mut candidates = entries.collect::<std::io::Result<Vec<_>>>()?;
    candidates.sort_by_key(std::fs::DirEntry::file_name);
    let mut fallback = None;
    for entry in candidates {
        let path = entry.path();
        if !matches!(
            path.extension().and_then(|value| value.to_str()),
            Some("yml" | "yaml")
        ) {
            continue;
        }
        let filename = entry.file_name().to_string_lossy().into_owned();
        let contents = fs::read_to_string(path)?;
        if contents.contains("npm publish") || contents.contains("npm stage publish") {
            return Ok(Some(filename));
        }
        if fallback.is_none() && filename.contains("release") {
            fallback = Some(filename);
        }
    }
    Ok(fallback)
}
